source/SAMRAI/solv/CellPoissonHypreSolver.C

```diff
@@ -682,11 +682,11 @@ CellPoissonHypreSolver::copyToHypre(
 {
    TBOX_ASSERT_DIM_OBJDIM_EQUALITY2(d_dim, src, box);
 
-   pdat::CellIterator cend(pdat::CellGeometry::end(box));
-   for (pdat::CellIterator c(pdat::CellGeometry::begin(box)); c != cend; ++c) {
-      hier::IntVector ic = *c;
-      HYPRE_StructVectorSetValues(vector, &ic[0], src(*c, depth));
-   }
+   pdat::CellData<double> tmp(box,1,hier::IntVector::getZero(d_dim));
```

Comment:
Do you need to create the temporary copy? This will incur extra malloc/free overhead which can be expensive on some platforms.

Reply:
I need an array with no ghost values to give to hypre, at least the way the hypre solver is setup now. The CellData 'src' has ghost values.

Reply:
I think we ought to keep both options available, as in some cases it will be better to take advantage of the HYPRE BoxValues functions, and in others we want to avoid the temporary copy.  I'll propose something based on what @jeanlucf22 has contributed.

Reply:
Actually, the function copyToHypre() is called twice. One time with the initial solution/guess, one time with the rhs. If the initial guess is 0 (as it is often the case), we may be able to avoid setting values completely. And since the rhs has no ghost values, the copy could be avoided in that second call.
This being said, the  changes I made reduced the cost of these functions from 'significant' to 'negligible' in my application.

Reply:
Is a HYPRE_StructVector initialized to zero by default, thus making the initial guess usable without any kind of setValues call?

Reply:
I believe it is. But in general, an application may reuse the solver many times, and even if it is 0 the first time, it will not be the case for the second time it is used. 
So maybe a flag should be passed in when calling CellPoissonHypreSolver::solveSystem(u,f) telling that "u" is 0, and in that case we would simply call HYPRE_StructVectorSetConstantValues(d_linear_sol,0.) to reset the hypre vector d_linear_sol to 0?